File: slack-reader/src/test/java/dev/number6/slackreader/dagger/RecordingSlackReaderAdaptor.java
package dev.number6.slackreader.dagger;

import dev.number6.slack.port.HttpPort;
import dev.number6.slackreader.adaptor.SlackReaderAdaptor;
import dev.number6.slackreader.model.Channel;
import dev.number6.slackreader.model.Message;
import com.amazonaws.services.lambda.runtime.LambdaLogger;

import java.time.LocalDate;
import java.util.*;
import java.util.stream.Collectors;

public class RecordingSlackReaderAdaptor extends SlackReaderAdaptor {

    private Collection<String> channelNames = new ArrayList<>();
    private Map<String, List<String>> channelMessages = new HashMap<>();

    public RecordingSlackReaderAdaptor(HttpPort client) {
        super(client);
    }

    @Override
    public Collection<Channel> getChannelList(LambdaLogger logger) {
        Collection<Channel> channels = super.getChannelList(logger);
        channelNames = channels.stream().map(Channel::getName).collect(Collectors.toList());
        return channels;
    }

    @Override
    public Collection<Message> getMessagesForChannelOnDate(Channel c, LocalDate date, LambdaLogger logger) {
        Collection<Message> messages = super.getMessagesForChannelOnDate(c, date, logger);
        channelMessages.put(c.getName(), messages.stream().map(Message::getText).collect(Collectors.toList()));
        return messages;
    }

    public List<String> getMessagesForChannelName(String channelName) {
        return channelMessages.getOrDefault(channelName, new ArrayList<>());
    }

    public Collection<String> getChannelNames() {
        return channelNames;
    }
}

File: slack-reader/src/test/kotlin/dev/number6/slackreader/dagger/RecordingSlackReaderAdaptor.kt
package dev.number6.slackreader.dagger

import com.amazonaws.services.lambda.runtime.LambdaLogger
import dev.number6.slack.adaptor.SlackClientAdaptor
import dev.number6.slackreader.adaptor.SlackReaderAdaptor
import dev.number6.slackreader.model.Channel
import dev.number6.slackreader.model.Message
import java.time.LocalDate

class RecordingSlackReaderAdaptor(client: SlackClientAdaptor) : SlackReaderAdaptor(client) {
    private var channelNames: List<String> = listOf()
    private val channelMessages: MutableMap<String?, List<String?>> = mutableMapOf()
    override fun getChannelList(logger: LambdaLogger): Collection<Channel> {
        val channels = super.getChannelList(logger)
        channelNames = channels.map { c -> c.name }
        return channels
    }

    override fun getMessagesForChannelOnDate(c: Channel, date: LocalDate, logger: LambdaLogger): Collection<Message> {
        val messages = super.getMessagesForChannelOnDate(c, date, logger)
        channelMessages[c.name] = messages.map { m -> m.text }
        return messages
    }

    fun getMessagesForChannelName(channelName: String): List<String?> {
        return channelMessages.getOrDefault(channelName, listOf())
    }

    fun getChannelNames(): List<String> {
        return channelNames
    }
}